File: StriverA2Z/Java/BinaryTrees/NodeAtDistanceK.java
import com.sun.source.tree.Tree;

import java.util.*;

public class NodeAtDistanceK {
    public static void main(String[] args) {
        Node root = new Node(3);
        root.left = new Node(5);
        root.right = new Node(1);
        root.left.left = new Node(6);
        root.left.right = new Node(2);
        root.left.right.left = new Node(7);
        root.left.right.right = new Node(4);
        root.right.left = new Node(0);
        root.right.right = new Node(8);

        List<Integer> ans = distanceK(root, root.left, 2);
        for (int i : ans) {
            System.out.println(i + " ");
        }
    }

    public static List<Integer> distanceK(Node root, Node target, int k) {
        Map<Node, Node> parentTrack = new HashMap<>();
        markParents(root, parentTrack);
        Queue<Node> q = new LinkedList<>();
        Map<Node, Boolean> visited = new HashMap<>();
        q.offer(target);
        visited.put(target, true);
        int cur = 0;
        while (!q.isEmpty()) {
            if (cur == k) break;
            cur++;
            int size = q.size();
            for (int i = 0; i < size; i++) {
                Node node = q.poll();
                if (node.left != null && visited.get(node.left) == null) {
                    q.offer(node.left);
                    visited.put(node.left, true);
                }
                if (node.right != null && visited.get(node.right) == null) {
                    q.offer(node.right);
                    visited.put(node.right, true);
                }
                if (parentTrack.get(node) != null && visited.get(parentTrack.get(node)) == null) {
                    q.offer(parentTrack.get(node));
                    visited.put(parentTrack.get(node), true);
                }
            }
        }
        List<Integer> result = new ArrayList<>();
        while (!q.isEmpty()) {
            Node t = q.poll();
            result.add(t.data);
        }
        return result;
    }

    public static void markParents(Node root, Map<Node, Node> parentTrack) {
        Queue<Node> q = new LinkedList<>();
        q.offer(root);
        while (!q.isEmpty()) {
            Node curr = q.poll();
            if (curr.left != null) {
                parentTrack.put(curr.left, curr);
                q.offer(curr.left);
            }
            if (curr.right != null) {
                parentTrack.put(curr.right, curr);
                q.offer(curr.right);
            }
        }
    }
}

File: StriverA2Z/JavaScript&TypeScript/src/BinaryTrees/nodeAtDistanceK.ts
import {BTNode} from "./BinaryTreeOperations";

function markParents(root: BTNode, parentTrack: Map<BTNode, BTNode>) {
	let q = [];
	q.push(root);
	while (q.length > 0) {
		let curr = q.shift();
		if (curr.left != null) {
			parentTrack.set(curr.left, curr);
			q.push(curr.left);
		}
		if (curr.right != null) {
			parentTrack.set(curr.right, curr);
			q.push(curr.right);
		}
	}
}

function distanceK(root: BTNode, target: BTNode, k: number) {
	let parentTrack: Map<BTNode, BTNode> = new Map();
	markParents(root, parentTrack);
	let q = [];
	let visited = new Map();
	q.push(target);
	visited.set(target, true);
	let cur = 0;
	while (q.length > 0) {
		if (cur === k) break;
		cur++;
		let size = q.length;
		for (let i = 0; i < size; i++) {
			let node = q.shift();
			if (node.left && !visited.has(node.left)) {
				q.push(node.left);
				visited.set(node.left, true);
			}
			if (node.right != null && !visited.has(node.right)) {
				q.push(node.right);
				visited.set(node.right, true);
			}
			if (parentTrack.get(node) != null && !visited.has(parentTrack.get(node))) {
				q.push(parentTrack.get(node));
				visited.set(parentTrack.get(node), true);
			}
		}
	}
	let result = [];
	while (q.length > 0) {
		let t = q.shift();
		result.push(t.data);
	}
	return result;
}

let root = new BTNode(3);
root.left = new BTNode(5);
root.right = new BTNode(1);
root.left.left = new BTNode(6);
root.left.right = new BTNode(2);
root.left.right.left = new BTNode(7);
root.left.right.right = new BTNode(4);
root.right.left = new BTNode(0);
root.right.right = new BTNode(8);

let ans = distanceK(root, root.left, 2);
console.log(ans)